File: python/promptsafe/detector.py
"""
PromptSafe — Core Detection Engine (Python)
Detects prompt injection and system prompt leakage using
regex pattern matching and keyword scoring.
"""

import re
import json
import os
from typing import List, Dict, Any, Optional


# ─── Paths ────────────────────────────────────────────────────────────────────

_PATTERNS_FILE = os.path.join(os.path.dirname(__file__), "patterns.json")


# ─── Globals ──────────────────────────────────────────────────────────────────

_PATTERNS_DB: Optional[Dict] = None
_COMPILED_PATTERNS: Optional[List[Dict]] = None


def _load_patterns() -> Dict:
    global _PATTERNS_DB
    if _PATTERNS_DB is None:
        with open(_PATTERNS_FILE, "r", encoding="utf-8") as f:
            _PATTERNS_DB = json.load(f)
    return _PATTERNS_DB


def _compile_patterns() -> List[Dict]:
    global _COMPILED_PATTERNS
    if _COMPILED_PATTERNS is None:
        db = _load_patterns()
        compiled = []
        for cat in db.get("categories", []):
            patterns = []
            for p in cat.get("patterns", []):
                try:
                    patterns.append(re.compile(p, re.IGNORECASE | re.DOTALL))
                except re.error:
                    pass  # skip invalid regex
            compiled.append({
                "name": cat["name"],
                "weight": cat.get("weight", 10),
                "patterns": patterns,
            })
        _COMPILED_PATTERNS = compiled
    return _COMPILED_PATTERNS


def invalidate_cache():
    """Force patterns to be reloaded on next scan (useful after addPattern)."""
    global _PATTERNS_DB, _COMPILED_PATTERNS
    _PATTERNS_DB = None
    _COMPILED_PATTERNS = None


# ─── Risk Level ───────────────────────────────────────────────────────────────

def _risk_level(score: int) -> str:
    if score >= 80:
        return "critical"
    if score >= 60:
        return "high"
    if score >= 40:
        return "medium"
    if score >= 20:
        return "low"
    return "safe"


# ─── Keyword Scoring ──────────────────────────────────────────────────────────

_KEYWORD_WEIGHTS = {
    "high": 15,
    "medium": 8,
    "low": 3,
}


def _keyword_score(text: str) -> int:
    db = _load_patterns()
    keywords_config = db.get("keywords", {})
    text_lower = text.lower()
    score = 0
    for tier, weight in _KEYWORD_WEIGHTS.items():
        for kw in keywords_config.get(tier, []):
            if kw.lower() in text_lower:
                score += weight
    return score


# ─── Main Scanner ─────────────────────────────────────────────────────────────

def scan(text: str, threshold: int = 35) -> Dict[str, Any]:
    """
    Scan ``text`` for prompt injection signals.

    Returns a dict with:
      - input        : the original text (truncated to 500 chars)
      - score        : integer 0-100
      - isSafe       : bool
      - riskLevel    : 'safe' | 'low' | 'medium' | 'high' | 'critical'
      - reasons      : list of human-readable trigger descriptions
      - matchedPatterns : list of category names that triggered
    """
    if not isinstance(text, str):
        raise TypeError(f"text must be str, got {type(text).__name__}")

    categories = _compile_patterns()
    score = 0
    reasons: List[str] = []
    matched: List[str] = []

    # ── Regex matching ───────────────────────────────────────────────────────
    for cat in categories:
        for rx in cat["patterns"]:
            m = rx.search(text)
            if m:
                score += cat["weight"]
                matched.append(cat["name"])
                reasons.append(
                    f"[{cat['name']}] matched: «{m.group(0)[:80]}»"
                )
                break  # one hit per category is enough

    # ── Keyword scoring ──────────────────────────────────────────────────────
    kw_score = _keyword_score(text)
    score += kw_score

    # ── Cap at 100 ───────────────────────────────────────────────────────────
    score = min(score, 100)

    is_safe = score < threshold

    return {
        "input": text[:500],
        "score": score,
        "isSafe": is_safe,
        "riskLevel": _risk_level(score),
        "reasons": reasons,
        "matchedPatterns": list(set(matched)),
    }


def add_pattern(category: str, pattern: str, weight: int = 15) -> None:
    """
    Add a custom regex pattern at runtime.

    Args:
        category: Category name (new or existing).
        pattern:  Python regex string (case-insensitive).
        weight:   Score contribution per hit (default 15).
    """
    db = _load_patterns()

    # Update raw DB
    for cat in db.get("categories", []):
        if cat["name"] == category:
            cat["patterns"].append(pattern)
            break
    else:
        db["categories"].append({
            "name": category,
            "weight": weight,
            "patterns": [pattern],
        })

    # Force recompile
    invalidate_cache()
    _PATTERNS_DB = db   # keep updated DB in memory without re-reading file

File: npm/src/detector.js
'use strict';

/**
 * PromptSafe — Detection Engine (Node.js)
 * Detects prompt injection using regex + keyword scoring.
 * Zero external dependencies.
 */

const fs = require('fs');
const path = require('path');

// ─── Patterns DB ─────────────────────────────────────────────────────────────

const PATTERNS_FILE = path.join(__dirname, '..', '..', 'python', 'promptsafe', 'patterns.json');

let _db = null;
let _compiled = null;

function loadDB() {
  if (!_db) {
    const raw = fs.readFileSync(PATTERNS_FILE, 'utf8');
    _db = JSON.parse(raw);
  }
  return _db;
}

// ─── Regex Compatibility ──────────────────────────────────────────────────────

/**
 * Remove Python-style inline flags like (?i) (?s) (?m) that are not supported
 * by the Node.js RegExp constructor (which already uses the 'gis' flags).
 */
function stripInlineFlags(pattern) {
  return pattern.replace(/\(\?[imsxuLs]+\)/g, '');
}

function compilePatterns() {
  if (!_compiled) {
    const db = loadDB();
    _compiled = (db.categories || []).map(cat => {
      const regexes = [];
      for (const p of (cat.patterns || [])) {
        try {
          regexes.push(new RegExp(stripInlineFlags(p), 'gis'));
        } catch (e) {
          // skip invalid patterns
        }
      }
      return { name: cat.name, weight: cat.weight || 10, regexes };
    });
  }
  return _compiled;
}

function invalidateCache() {
  _db = null;
  _compiled = null;
}

// ─── Risk Level ───────────────────────────────────────────────────────────────

function riskLevel(score) {
  if (score >= 80) return 'critical';
  if (score >= 60) return 'high';
  if (score >= 40) return 'medium';
  if (score >= 20) return 'low';
  return 'safe';
}

// ─── Keyword Scoring ─────────────────────────────────────────────────────────

const KEYWORD_WEIGHTS = { high: 15, medium: 8, low: 3 };

function keywordScore(text) {
  const db = loadDB();
  const keywords = db.keywords || {};
  const lc = text.toLowerCase();
  let score = 0;
  for (const [tier, weight] of Object.entries(KEYWORD_WEIGHTS)) {
    for (const kw of (keywords[tier] || [])) {
      if (lc.includes(kw.toLowerCase())) score += weight;
    }
  }
  return score;
}

// ─── Main Scanner ─────────────────────────────────────────────────────────────

/**
 * Scan text for prompt injection.
 *
 * @param {string} text
 * @param {Object} [opts]
 * @param {number} [opts.threshold=35]
 * @returns {{ input, score, isSafe, riskLevel, reasons, matchedPatterns }}
 */
function scan(text, opts = {}) {
  if (typeof text !== 'string') throw new TypeError('text must be a string');

  const threshold = opts.threshold !== undefined ? opts.threshold : 35;
  const categories = compilePatterns();

  let score = 0;
  const reasons = [];
  const matched = new Set();

  // Regex matching
  for (const cat of categories) {
    for (const rx of cat.regexes) {
      rx.lastIndex = 0; // reset for 'g' flag
      const m = rx.exec(text);
      if (m) {
        score += cat.weight;
        matched.add(cat.name);
        reasons.push(`[${cat.name}] matched: «${m[0].slice(0, 80)}»`);
        break; // one hit per category
      }
    }
  }

  // Keyword scoring
  score += keywordScore(text);

  // Cap at 100
  score = Math.min(score, 100);

  return {
    input: text.slice(0, 500),
    score,
    isSafe: score < threshold,
    riskLevel: riskLevel(score),
    reasons,
    matchedPatterns: [...matched],
  };
}

/**
 * Add a custom regex pattern at runtime.
 *
 * @param {string} category - Category name (new or existing).
 * @param {string} pattern  - JS-compatible regex string.
 * @param {number} [weight=15]
 */
function addPattern(category, pattern, weight = 15) {
  const db = loadDB();
  const existing = (db.categories || []).find(c => c.name === category);
  if (existing) {
    existing.patterns.push(pattern);
  } else {
    db.categories.push({ name: category, weight, patterns: [pattern] });
  }
  invalidateCache();
  _db = db; // keep updated DB without re-reading file
}

module.exports = { scan, addPattern };
